options/posix/generic/pthread-stubs.cpp

```diff
@@ -927,15 +927,77 @@ int pthread_cond_wait(pthread_cond_t *__restrict cond, pthread_mutex_t *__restri
 
 int pthread_cond_timedwait(pthread_cond_t *__restrict cond, pthread_mutex_t *__restrict mutex,
 		const struct timespec *__restrict abstime) {
+	constexpr long nanos_per_second = 1'000'000'000;
+	if (abstime && (abstime->tv_nsec < 0 || abstime->tv_nsec >= nanos_per_second))
+		return EINVAL;
+
 	auto seq = __atomic_load_n(&cond->__mlibc_seq, __ATOMIC_ACQUIRE);
-	// TODO: do proper error handling here.
-	if(pthread_mutex_unlock(mutex))
-		__ensure(!"Failed to unlock the mutex");
-	if(mlibc::sys_futex_wait(&cond->__mlibc_seq, seq, abstime))
-		__ensure(!"sys_futex_wait() failed");
-	if(pthread_mutex_lock(mutex))
-		__ensure(!"Failed to lock the mutex");
-	return 0;
+
+	// TODO: handle locking errors and cancellation properly.
+	while (true) {
+		if (pthread_mutex_unlock(mutex))
+			__ensure(!"Failed to unlock the mutex");
+
+		int e;
+		if (abstime) {
+			// Adjust for the fact that sys_futex_wait accepts a *timeout*, but
+			// pthread_cond_timedwait accepts an *absolute time*.
+			// Note: mlibc::sys_clock_get is available unconditionally.
+			// TODO: allow clock to be configured via a condattr.
+			struct timespec now;
+			if (mlibc::sys_clock_get(CLOCK_REALTIME, &now.tv_sec, &now.tv_nsec))
+				__ensure(!"sys_clock_get() failed");
+
+			struct timespec timeout;
+			timeout.tv_sec = abstime->tv_sec - now.tv_sec;
+			timeout.tv_nsec = abstime->tv_nsec - now.tv_nsec;
+
+			// Check if abstime has already passed.
+			if (timeout.tv_sec < 0 || (timeout.tv_sec == 0 && timeout.tv_nsec < 0)) {
+				if (pthread_mutex_lock(mutex))
+					__ensure(!"Failed to lock the mutex");
+				return ETIMEDOUT;
+			} else if (timeout.tv_nsec >= nanos_per_second) {
+				timeout.tv_nsec -= nanos_per_second;
```

Comment:
Couldn't ``tv_nsec`` be multiple times larger than ``nanos_per_second``?

Reply:
I don't think so, since both abstime.tv_nsec and now.tv_nsec should be less than nanos_per_sec, but I'll add an ensure here why not.

Reply:
That makes sense, it would be good to add an ensure in the function at least.

Reply:
POSIX says that tv_nsec is in range [0, 1000 million) (and any set/sleep function will fail with EINVAL if it's not for example).